### PR TITLE
Generate *Ext traits by default and allow disabling them instead

### PR DESCRIPTION
**Pull request by sdroege**
WIP: The configuration should probably be renamed to allow_no_trait or
similar instead.

Reason for this is that the general case is that subclasses of any
GObject are allowed, and in your own crate you don't generally know if
other crates or any other code is going to create subclasses. Only in
few cases a class is "final" and does not allow subclasses.

@EPashkin Any opinions on this one?

### CODE REVIEW

**Comment by EPashkin** [comment]
@sdroege Reasonable, can you give examples of final classes?

**Comment by EPashkin** [comment]
If this accepted `README.md` and object analyze  need changed too.

**Comment by sdroege** [comment]
> @sdroege Reasonable, can you give examples of final classes?

Everything coming from G_DECLARE_FINAL_TYPE, and every type where in C the struct definitions are not inside the public headers. It's nothing that we can detect easly

> If this accepted README.md and object analyze need changed too.

Sure :) Would you prefer the configuration to be renamed to something more sensible though? allow_no_trait would seem like a good candidate

**Comment by EPashkin** [comment]
Seems it really almost no final classes.
I prefer `trait=false` with renaming config variable `force_trait` to `generate_trait`

**Comment by sdroege** [comment]
Ok, will update accordingly later

**Comment by EPashkin** [comment]
This PR caused strange build error in gtk/auto/css_provider.rs 
```
   Compiling gtk v0.1.3 (file:///D:/eap/rust/0/gtk)
error[E0034]: multiple applicable items in scope
  --> D:/eap/rust/0/gtk\src\auto\css_provider.rs:55:30
   |
55 |         write!(f, "{}", self.to_string())
   |                              ^^^^^^^^^ multiple `to_string` found
   |
note: candidate #1 is defined in an impl of the trait `auto::css_provider::CssProviderExt` for the type `_`
  --> D:/eap/rust/0/gtk\src\auto\css_provider.rs:98:5
   |
98 | /     fn to_string(&self) -> String {
99 | |         unsafe {
100| |             from_glib_full(ffi::gtk_css_provider_to_string(self.to_glib_none().0))
101| |         }
102| |     }
   | |_____^
   = note: candidate #2 is defined in an impl of the trait `std::string::ToString` for the type `_`
```

**Comment by EPashkin** [comment]
Also it generated trait when no trait members ex for `gtk::DrawingAreaExt`
And causes strange error in https://github.com/gtk-rs/examples/blob/master/src/bin/menu_bar.rs despite trait must be included with `use gtk::prelude::*;`
```
   Compiling gtk-rs-examples v0.0.1 (file:///D:/eap/rust/0/examples)
error: no method named `set_submenu` found for type `gtk::MenuItem` in the current scope
  --> src\bin\menu_bar.rs:58:10
   |
58 |     file.set_submenu(Some(&menu));
   |          ^^^^^^^^^^^
   |
   = help: items from traits can only be used if the trait is in scope; the following trait is implemented but not in scope, perhaps add a `use` for it:
   = help: candidate #1: `use gtk::MenuItemExt;`
```

**Comment by sdroege** [comment]
I'll look at the CSS thing.

For the menubar trait, maybe the traits should all be directly re-exported from the lib.rs? Not sure why the trait is needed there though, and empty traits should probably be just skipped. I'll take a look

**Comment by sdroege** [comment]
The CSS trait confusion is solved now, empty traits are not generated anymore, and the example you linked above also compiles for me now. Please review again :)

**Comment by sdroege** [comment]
One problem here seems to be that now almost everything goes into the Ext trait, but we can't manually implement any of those functions of the Ext trait. One would have manually write an ExtManual trait for those things.

I guess not really a problem in the end, just slightly ugly, and it's certainly not affecting existing code.

**Comment by EPashkin** [comment]
Now has problem after generating gio app_launch_context.rs and gdk drag_context.rs, both has `use glib;` under feature and examples don't builds.

I still get 6 errors in menu_bar after fix `use glib` manually

**Comment by sdroege** [comment]
I'll take a look at gio later. How can I reproduce the menu bar problem?

**Comment by EPashkin** [comment]
I build pending branch of examples with using http://doc.crates.io/specifying-dependencies.html#overriding-with-local-dependencies to replace all gtk-rs crates.
When replacing only gtk it compiles.
Trying to find what crate cause problem.

**Comment by EPashkin** [comment]
Problem solved: gio also have `MenuItemExt` and we rexport it in gtk::prelude (not sure why).

File: src/analysis/object.rs [hunk 1]
```diff
@@ -104,7 +104,7 @@ pub fn class(env: &Env, obj: &GObject, deps: &[library::TypeId]) -> Option<Info>
     let child_properties = child_properties::analyze(env, obj.child_properties.as_ref(), class_tid,
                                                      &mut imports);
 
-    if has_children && !properties.is_empty() {
+    if generate_trait && !properties.is_empty() {
```
Comment (EPashkin) [inline src/analysis/object.rs]: `||` here, seems we found old bug :wink:
It source bad gio and gdk generation.

Comment (EPashkin) [inline src/analysis/object.rs]: Now unneeded using in gio simple_permision.rs

Comment (EPashkin) [inline src/analysis/object.rs]: Seems this need changed to with removing lines 82-84
```
    let has_methods = functions.iter()
            .any(|f| f.kind == library::FunctionKind::Method);
    if (generate_trait && has_methods) || !properties.is_empty() {
        imports.add("glib", None);
        imports.add("glib::object::IsA", None);
    }
```

File: src/analysis/object.rs [hunk 2]
```diff
@@ -66,20 +66,20 @@ pub fn class(env: &Env, obj: &GObject, deps: &[library::TypeId]) -> Option<Info>
...
         let child_name = child_tid.full_name(&env.library);
         let status = env.config.objects.get(&child_name)
             .map(|o| o.status)
             .unwrap_or_default();
         if status.normal() {
-            has_children = true;
+            generate_trait = true;
             break;
         }
     }
```
Comment (EPashkin) [inline src/analysis/object.rs]: This for unneeded

File: src/codegen/doc/mod.rs [hunk 3]
```diff
@@ -91,7 +91,7 @@ fn create_object_doc(w: &mut Write, env: &Env, info: &analysis::object::Info) ->
     let symbols = env.symbols.borrow();
     let ty = TypeStruct::new(SType::Struct, &info.name);
     let ty_ext = TypeStruct::new(SType::Trait, &format!("{}Ext", info.name));
-    let has_trait = info.has_children;
+    let has_trait = info.generate_trait;
```
Comment (EPashkin) [inline src/codegen/doc/mod.rs]: Maybe something like `generate_trait(info)`

File: src/analysis/object.rs (hunk 1 above)
Comment (EPashkin) [inline src/analysis/object.rs]: Or better as you done in `generate_trait` function

**Comment by EPashkin** [comment]
@GuillaumeGomez you don't remember reason for [this](https://github.com/gtk-rs/gtk/blob/master/src/prelude.rs#L8) reexport?

**Comment by sdroege** [comment]
Thanks, I'll update the pull request tomorrow :) what other changes did you do elsewhere to make things work?

File: src/analysis/object.rs (hunk 1 above)
Comment (EPashkin) [inline src/analysis/object.rs]: Unoptimized version that give needed result:
```
    let has_methods = functions.iter()
            .any(|f| f.kind == library::FunctionKind::Method);
    let has_signals = signals.iter().any(|s| s.trampoline_name.is_ok());
    if generate_trait && (!properties.is_empty() || !child_properties.is_empty() || has_signals) {
        imports.add("glib", None);
    }
    if generate_trait && (has_methods || !properties.is_empty() || !child_properties.is_empty() || has_signals) {
        imports.add("glib::object::IsA", None);
    }
```

**Comment by EPashkin** [comment]
In gir I only changed imports, on gtk removed gio reexport.
After this all generated code compiles without additional warnings and look normal at first glance.

File: src/analysis/object.rs (hunk 1 above)
Comment (EPashkin) [inline src/analysis/object.rs]: First if don't need `child_properties`

**Comment by EPashkin** [comment]
Affected crates: gio, pango, gdk, gtk

**Comment by GuillaumeGomez** [comment]
> @GuillaumeGomez you don't remember reason for [this](https://github.com/gtk-rs/gtk/blob/master/src/prelude.rs#L8) reexport?

Not at all. We should try to remove it and see what's affected.

**Comment by EPashkin** [comment]
Examples builds fine.

**Comment by EPashkin** [comment]
After some fix and last check ;)

File: src/analysis/object.rs (hunk 2 above)
Comment (sdroege) [inline src/analysis/object.rs]: Why? If obj.generate_trait=false (i.e. set false in the configuration), we would still want to generate a trait if it is actually needed. Or not?

**Comment by sdroege** [comment]
Alright, thanks for the review and checking things :) I'll update the pull request with your comments and then let's check everything again.

For future reference, which crates are you all testing with any changes to the code generator? Only those under the gtk-rs project, or also others? And which testcases do you compile/run?

File: src/analysis/object.rs (hunk 2 above)
Comment (EPashkin) [inline src/analysis/object.rs]: I don't think that Gir must override human decision about generating trait,
at maximum in can produce error! to inform.

**Comment by EPashkin** [comment]
I check meaningful changes for all crates under gtk-rs
and try build examples pending branch for default feature and gtk_3_22 (almost as CI will do later).

File: src/analysis/object.rs (hunk 2 above)
Comment (sdroege) [inline src/analysis/object.rs]: OK, error! it is then :)

Comment (EPashkin) [inline src/analysis/object.rs]: And separate for to function please.

**Comment by sdroege** [comment]
Please see attached. I hope I got all your review comments correctly. Note that I just calculate whether a trait is to be generated once in the analysis, instead of duplicating all these conditions in the code generator and docs.

File: src/analysis/object.rs [hunk 4]
```diff
@@ -66,23 +66,23 @@ pub fn class(env: &Env, obj: &GObject, deps: &[library::TypeId]) -> Option<Info>
...
+    if !generate_trait {
+        for child_tid in env.class_hierarchy.subtypes(class_tid) {
+            let child_name = child_tid.full_name(&env.library);
+            let status = env.config.objects.get(&child_name)
+                .map(|o| o.status)
+                .unwrap_or_default();
+            if status.normal() {
+                error!("Not generating trait for {} although subtypes exist", full_name);
+                break;
+            }
```
Comment (EPashkin) [inline src/analysis/object.rs]: IMHO this better separate function like `has_subtypes` here.
But I don't insist this

File: src/codegen/trait_impls.rs [hunk 5]
```diff
@@ -28,17 +28,25 @@ fn lookup<'a>(functions: &'a [Info], name: &str) -> &'a Info {
         .unwrap()
 }
 
-fn generate_display(w: &mut Write, type_name: &str, func: &Info) -> Result<()> {
+fn generate_display(w: &mut Write, type_name: &str, func: &Info, in_trait: bool) -> Result<()> {
     use analysis::out_parameters::Mode;
+
```
Comment (EPashkin) [inline src/codegen/trait_impls.rs]: Unneeded empty line

File: src/codegen/trait_impls.rs [hunk 6]
```diff
@@ -49,43 +57,67 @@ impl fmt::Display for {type_name} {{
...
 
-fn generate_ord(w: &mut Write, type_name: &str, func: &Info) -> Result<()> {
+fn generate_ord(w: &mut Write, type_name: &str, func: &Info, in_trait: bool) -> Result<()> {
+    let call = if in_trait {
+        format!("{type_name}Ext::{func_name}(self, other)",
+                type_name = type_name, func_name = func.name)
+    } else {
+        format!("self.{func_name}(other)",
+                func_name = func.name)
+    };
```
Comment (EPashkin) [inline src/codegen/trait_impls.rs]: This block repeated many times, maybe better move it to function?

**Comment by sdroege** [comment]
Next try :)

File: src/codegen/trait_impls.rs [hunk 7]
```diff
@@ -28,18 +28,49 @@ fn lookup<'a>(functions: &'a [Info], name: &str) -> &'a Info {
...
+        if in_trait {
+            args_string.push_str(", ");
+        }
+        for arg in args {
+            args_string.push_str(arg);
+            args_string.push_str(", ");
+        }
+        // Drop final ", "
+        let len = args_string.len();
+        args_string.truncate(len - 2);
```
Comment (EPashkin) [inline src/codegen/trait_impls.rs]: args_string.push_str(&args.join(", ")); ?

Comment (sdroege) [inline src/codegen/trait_impls.rs]: Done, good point

**Comment by EPashkin** [comment]
@sdroege Thanks.
I don't fully understand usability of reviews, but IMHO you now have `Merge pull request` button.
Please do merge when CI passed.

**Comment by sdroege** [comment]
There's no such button for me :) It says
>  Only those with write access to this repository can merge pull requests. 

Thanks for the review

**Comment by EPashkin** [comment]
Bad, then I merge it

**Comment by sdroege** [comment]
Thanks. I'll send pull requests for updating the other crates tomorrow.

Working on the free functions generation currently

**Pull request merged**